cpp/include/cudf/join/sort_merge_join.hpp

```diff
@@ -0,0 +1,283 @@
+                          ///< raw table view if null equality is set to false, and is equal to the
+                          ///< raw table view otherwise
+
+    std::optional<rmm::device_buffer> raw_validity_mask =
+      std::nullopt;  ///< optional filters for null_equality::UNEQUAL
+    std::optional<size_type> raw_num_nulls =
+      std::nullopt;  ///< optional filters for null_equality::UNEQUAL
+    std::optional<std::unique_ptr<table>> tbl =
+      std::nullopt;  ///< optional filters for null_equality::UNEQUAL
+
+    std::optional<std::unique_ptr<column>> tbl_sorted_order =
+      std::nullopt;  ///< optional reordering if we are given pre-sorted tables
+
+    /**
+     * @brief Mark rows in raw table with nulls at root or child levels by populating the
+     * raw_validity_mask
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     * @param mr Device memory resource used to allocate the table and columns' device
+     * memory.
+     */
+    void populate_nonnull_filter(rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+    /**
+     * @brief Apply raw_validity_mask on the raw_tbl_view to create a null-free table
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     * @param mr Device memory resource used to allocate the table and columns' device
+     * memory.
+     */
+    void apply_nonnull_filter(rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+    /**
+     * @brief Pre-process the raw table in the case where nulls are unequal.
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     * @param mr Device memory resource used to allocate the table and columns' device
+     * memory.
+     */
+    void preprocess_raw_table(rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+    /**
+     * @brief Get sorted ordering of processed table
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     * @param mr Device memory resource used to allocate the table and columns' device
+     * memory.
+     */
+    void get_sorted_order(rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+    /**
+     * @brief Get mapping from processed table to raw table to return correct join indices
+     *
+     * @param stream CUDA stream used for device memory operations and kernel launches
+     * @param mr Device memory resource used to allocate the table and columns' device
+     * memory.
+     * @return A device vector indicating the mapping between pre-processed table and raw table
+     */
+    rmm::device_uvector<size_type> map_tbl_to_raw(rmm::cuda_stream_view stream,
+                                                  rmm::device_async_resource_ref mr);
+  };
+
+ private:
+  preprocessed_table ptleft;
+  preprocessed_table ptright;
```

Comment:
```suggestion
  preprocessed_table preprocessed_left;
  preprocessed_table preprocessed_right;
```
nit to improve readability